Prepare for Unicode 14

Manishearth opened the pull request:
This PR:

 - Adds Unicode 14 to all the requisite enums, etc
 - Copies Unicode 13's `ucd` and `emoji` folders.
 - Applies updated version/date headers to these files

(No other changes are made to the data files)

Some things missing from the [documentation](https://sites.google.com/site/unicodetools/home):
 - The `.bat` files are no longer generated and I couldn't figure out how to do it. Instead I followed the instructions [here](https://sites.google.com/site/unicodetools/inputdata) with the Python script
 - The `extra` and `cldr` folders do not exist under `unicodetools/data/ucd/14.0.0`, but they do under `Generated`. I only updated files which existed
 - Blocks.txt is under `unicodetools/data/ucd/14.0.0` in unicodetools, and `Generated/UCD/d1/extra` in Generated. Copying it out is not enough; the Generated Blocks.txt uses short names for blocks, whereas the unicodetools one uses long ones, and UCD asks for an additional file for the mapping. For now I just patched the header on Blocks.txt, but it seems like we should be generating `d1/Blocks.txt` as well with the long codes.

I can upload the `d1` files if you tell me when and where.

cc @markusicu @Ken-Whistler

macchiati reviewed (commented):
What I usually do is diff them against the previous version, to check that the data looks sensible. I can do that, but probably later during the week.

One quick note is that the emoji files are showing version 13.0, which will need fixing. Not your problem since you were going by the instructions (and they were clearly incomplete)

Manishearth commented:
> One quick note is that the emoji files are showing version 13.0, which will need fixing. Not your problem since you were going by the instructions (and they were clearly incomplete)

Should be easy enough to fix.

markusicu commented on unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt:
```diff
@@ -1,6 +1,6 @@
-Generate: .
-CopyrightYear: 2020
-DeltaVersion: 17
+Generate: .*
```
What does the star do?

Manishearth replied:
"Generate everything", i think this was in the docs somewhere, but I can revert it

markusicu commented on unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt:
```diff
@@ -225,6 +225,9 @@ Value:  V12_1
 # Newly assigned in Unicode 13.0.0 (March, 2020)
 
+Value:  V14_0
+# Newly assigned in Unicode 14.0.0 (?, 2021)
```
September

http://www.unicode.org/versions/Unicode14.0.0/

markusicu commented on unicodetools/org/unicode/text/utility/Utility.java:
```diff
@@ -826,6 +826,7 @@ public static String join(long[] array, String divider) {
 
     public static final String[] searchPath = {
 	    // "EXTRAS" + (FIX_FOR_NEW_VERSION == 0 ? "" : ""),
+    	"14.0.0",
```
indentation? tabs?

markusicu reviewed (commented):
Thanks, looks plausible for a start.

I always do a combination of following the notes on the unicodetools site plus “forensic programming”. I don't remember the details from one year to the next :-}

I see that you mostly copied Unicode 13 files as-is, but also put some Unicode 14 files in here. I think I used to start with a commit of just the old files in the new folders, for a more visible file history. Not sure if it makes much of a difference though. It *might* help git with the file history if the initial diffs for the new version are large enough to exceed git's threshold for whether the new version looks similar to the old one and record a "copy". Probably less important generally than in svn.

We will also need to copy the version-13 files for idna, security, and UCA, for a clean start there. Can be a separate PR.

As for generated files, they don't go directly back into the repo. We post the generated files to a location where KenW grabs them, looks them over, and if they are ok, he posts them in the Unicode Public folder. Then we grab all of the files from there again, put them into the repo's data folders, and run the tools again. Etc.

For example, here is where I posted files that I generated for Unicode 13: https://corp.unicode.org/~book/incoming/markus/u13/ plus (once UCD is settled) https://corp.unicode.org/~book/incoming/markus/uca13/

markusicu commented:
> * The `extra` and `cldr` folders do not exist under `unicodetools/data/ucd/14.0.0`, but they do under `Generated`. I only updated files which existed

That's right. They are files for diagnosis or for use in CLDR and ICU, but don't go into the UCD/UCA/...

To be clear, I assume that what you put into the data input folders is straight from https://www.unicode.org/Public/14.0.0/ right? Do not check in things from the Generated output folder. The generated files first go for review, and we only pick them up after KenW has looked them over and copied them to /Public.

> * Blocks.txt is under `unicodetools/data/ucd/14.0.0` in unicodetools, and `Generated/UCD/d1/extra` in Generated. Copying it out is not enough; the Generated Blocks.txt uses short names for blocks, whereas the unicodetools one uses long ones, and UCD asks for an additional file for the mapping. For now I just patched the header on Blocks.txt, but it seems like we should be generating `d1/Blocks.txt` as well with the long codes.

I think we update Blocks.txt manually, post for Ken, wait for his approval and copy to /Public, and ony then copy into the data folder.

Manishearth commented:
> To be clear, I assume that what you put into the data input folders is straight from https://www.unicode.org/Public/14.0.0/ right? Do not check in things from the Generated output folder. The generated files first go for review, and we only pick them up after KenW has looked them over and copied them to /Public.

No, I copied them from Unicode 13, as listed in the docs. I can copy those files instead.

Manishearth commented:
(once this is done I hope to update the docs with everything I've learned)

markusicu commented:
> > To be clear, I assume that what you put into the data input folders is straight from https://www.unicode.org/Public/14.0.0/ right? Do not check in things from the Generated output folder. The generated files first go for review, and we only pick them up after KenW has looked them over and copied them to /Public.
> 
> No, I copied them from Unicode 13, as listed in the docs. I can copy those files instead.

You copied the ucd and emoji files from the data/.../13.0.0 folders to the data/.../14.0.0 folders, but you also then overwrote some files like unicodetools/data/ucd/14.0.0-Update/PropertyAliases.txt which has a version 14 header. What I am saying is that these files should only ever come from the /Public folder, never directly from the Generated output.

Manishearth commented:
Ah. I'll make an additional commit copying in 14 files from Public

Manishearth commented:
> but you also then overwrote some files like unicodetools/data/ucd/14.0.0-Update/PropertyAliases.txt which has a version 14 header

Oh, uh, not quite, this is the Unicode 13 file, I just added the Unicode 14 header to make the tools stop complaining about diffs 😄

Manishearth commented:
- Copied in the generated files
 - Addressed inline comments
 - Added new scripts and blocks to the code. I did _not_ add them to the data files (e.g. PropertyValueAliases.txt for short codes), I expect they will be autogenerated there, but I can add them if necessary)

Currently stuck on

```
Exception in thread "main" java.lang.IllegalArgumentException: Failed to find Script value for 1E290
	at org.unicode.text.UCD.ToolUnicodePropertySource$ToolUnicodeProperty._getValue(ToolUnicodePropertySource.java:1653)
```

but 1E290 has an entry in Scripts.txt, so I'm not sure why it's not getting loaded

Manishearth commented:
It seems to be generating again, after working through all the errors. Note for when I document stuff; we need to document "new scripts" and "new blocks".

Manishearth commented:
Oh, another thing:

 - The emoji/, auxiliary/, and extracted/ folders on /Public are empty. I didn't touch them
 - Unihan.zip does not contain files that overlap with Unihan/, so I didn't update that

The build throws these non-blocking errors:

```
***Rules missing from SAMPLES for Line: [12.3, 13.04, 21.2, 27.01]
...
java.lang.IllegalArgumentException: Internal error: cjkAccountingNumeric doesn't contain : []
java.lang.IllegalArgumentException: Internal error: cjkOtherNumeric doesn't contain : []
java.lang.IllegalArgumentException: Internal error: cjkPrimaryNumeric doesn't contain : []
```

markusicu commented:
> Oh, uh, not quite, this is the Unicode 13 file, I just added the Unicode 14 header to make the tools stop complaining about diffs 😄

I wouldn't make this kind of edit. The data files should be the old versions, or come from the /Public online folder, or have real edits for new data.

I *believe* that I used to make one commit (in svn) with just copies within the repo from the old-version data/ files to the new-version data/ files, and subsequent commits where I copied in /Public files and/or added new data.

> Copied in the generated files

This sounds wrong, as discussed...

> The emoji/, auxiliary/, and extracted/ folders on /Public are empty. I didn't touch them

They are empty until the files we generated get reviewed and posted.

> Unihan.zip does not contain files that overlap with Unihan/, so I didn't update that

The files in Unihan.zip are hard to work with. Hard to diff, and properties can move between files.
Instead, we re-munge them into one file per property.
Instructions are in the "Unihan" section of https://sites.google.com/site/unicodetools/inputdata
(We will need to change that slightly to talk about git instead of svn.)

----
FYI  I expect to post the last two ISO script codes tomorrow that we need for Unicode 14.
(I am the ISO 15924 registrar. There is a whole process for that.)

Manishearth commented:
> I wouldn't make this kind of edit. The data files should be the old versions, or come from the /Public online folder, or have real edits for new data.


Yep, the files are now from /Public

Manishearth commented:
Well, except for the emoji files, which don't exist and cause errors

Manishearth commented:
Okay, the current state is that there are no more "copied from Unicode 13" files in `14.0.0-Update`. 

There are still files under `emoji/14.0` that are copied from Unicode 13, because the program fails without it. What should my next steps be here: should I be getting Ken to upload my generated files? Or getting this PR merged?

markusicu commented on unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt:
```diff
@@ -1,6 +1,6 @@
-CopyrightYear: 2020
-DeltaVersion: 17
+CopyrightYear: 2021
+DeltaVersion: 1
```
We usually set this one higher than the highest dsomething that is published, so that there is no collision. Looks like we have a d7 now, so this wants to be 8 so that your generated files say d8.

markusicu commented on unicodetools/org/unicode/text/UCD/ShortBlockNames.txt:
```diff
@@ -291,19 +299,23 @@ Thai	;	Thai
 Yi_Syllables	;	Yi_Syllables
 Yijing	;	Yijing_Hexagram_Symbols
 Zanabazar_Square	;	Zanabazar_Square
+Znamenny_Musical ; Znamenny_Musical_Notation
```
```suggestion
Znamenny_Music ; Znamenny_Musical_Notation
```
see https://www.unicode.org/Public/14.0.0/ucd/PropertyValueAliases-14.0.0d2.txt

markusicu reviewed (commented):
FYI  Mark has a competing PR now adding emoji 13.1. Let's see who gets in first vs. who gets to merge :-)

FYI  Next PR could update to https://www.unicode.org/Public/14.0.0/ucd/PropertyValueAliases-14.0.0d2.txt which adds new block and script names.

macchiati commented:
Manish started first; the 13.1 update isn't as important...

Mark


On Mon, Mar 1, 2021 at 1:11 PM Markus Scherer <notifications@github.com>
wrote:

> *@markusicu* commented on this pull request.
>
> FYI Mark has a competing PR now adding emoji 13.1. Let's see who gets in
> first vs. who gets to merge :-)
>
> FYI Next PR could update to
> https://www.unicode.org/Public/14.0.0/ucd/PropertyValueAliases-14.0.0d2.txt
> which adds new block and script names.
> ------------------------------
>
> In unicodetools/data/emoji/14.0/ReadMe.txt
> <https://github.com/unicode-org/unicodetools/pull/40#discussion_r585043688>
> :
>
> > @@ -0,0 +1,21 @@
>
> +# Unicode Emoji
>
> +# © 2020 Unicode®, Inc.
>
> +# Unicode and the Unicode Logo are registered trademarks of Unicode, Inc. in the U.S. and other countries.
>
> +# For terms of use, see http://www.unicode.org/terms_of_use.html
>
> +
>
> +This directory contains data files for Unicode Emoji, Version 13.0
>
> +
>
> +Public/emoji/13.0/
>
>
> The sequences files should probably start with 13.1 not 13.0:
> https://www.unicode.org/Public/emoji/13.1/
> ------------------------------
>
> In unicodetools/org/unicode/text/UCD/MakeUnicodeFiles.txt
> <https://github.com/unicode-org/unicodetools/pull/40#discussion_r585049369>
> :
>
> > @@ -1,6 +1,6 @@
>
>  Generate: .
>
> -CopyrightYear: 2020
>
> -DeltaVersion: 17
>
> +CopyrightYear: 2021
>
> +DeltaVersion: 1
>
>
> We usually set this one higher than the highest dsomething that is
> published, so that there is no collision. Looks like we have a d7 now, so
> this wants to be 8 so that your generated files say d8.
> ------------------------------
>
> In unicodetools/org/unicode/text/UCD/ShortBlockNames.txt
> <https://github.com/unicode-org/unicodetools/pull/40#discussion_r585050024>
> :
>
> >  Wancho	;	Wancho
>
>  Warang_Citi	;	Warang_Citi
>
>  Yezidi	;	Yezidi
>
>  Yi_Radicals	;	Yi_Radicals
>
>  Yi_Syllables	;	Yi_Syllables
>
>  Yijing	;	Yijing_Hexagram_Symbols
>
>  Zanabazar_Square	;	Zanabazar_Square
>
> +Znamenny_Musical ; Znamenny_Musical_Notation
>
>
> ⬇️ Suggested change
>
> -Znamenny_Musical ; Znamenny_Musical_Notation
>
> +Znamenny_Music ; Znamenny_Musical_Notation
>
>
> see
> https://www.unicode.org/Public/14.0.0/ucd/PropertyValueAliases-14.0.0d2.txt
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/unicode-org/unicodetools/pull/40#pullrequestreview-601158623>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACJLEMB4DFKPEQXN6DRWTCDTBP7I5ANCNFSM4XOILQEA>
> .
>

Manishearth commented:
Should I still pull in 13.1? 

Also, from the ucd-dev list it seems like a lot of data files have been updated: Should I be pulling in the latest files before we can merge this?

macchiati commented:
13.1 was an emoji-sequences-only release, and the UCD files are identical
to 13.0.

 I just copied in 13.0 files because it makes it easier to build for the
UnicodeJsps (the tooling isn't set up quite right for having a delta like
that), so I think you can simply proceed with 14.0 as you were.

Mark


On Mon, Mar 1, 2021 at 1:35 PM Manish Goregaokar <notifications@github.com>
wrote:

> Should I still pull in 13.1?
>
> Also, from the ucd-dev list it seems like a lot of data files have been
> updated: Should I be pulling in the latest files before we can merge this?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/unicode-org/unicodetools/pull/40#issuecomment-788313161>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACJLEMFK6XANU23FGYLB6W3TBQCCBANCNFSM4XOILQEA>
> .
>

markusicu commented:
> Should I still pull in 13.1?

I think yes. Just the four files from https://www.unicode.org/Public/emoji/13.1/

> Also, from the ucd-dev list it seems like a lot of data files have been updated: Should I be pulling in the latest files before we can merge this?

Possible, but I suggest to do that in the next PR.

Manishearth commented:
> I think yes. Just the four files from https://www.unicode.org/Public/emoji/13.1/

Done (and addressed the other comments).

> Possible, but I suggest to do that in the next PR.

Will do, and then I suppose I should send the updated generated files to Ken?

markusicu commented:
> > Possible, but I suggest to do that in the next PR.
> 
> Will do, and then I suppose I should send the updated generated files to Ken?

Sure. Best is to post the generated files somewhere and send to Ken with cc to ucd-dev.

We usually post to https://corp.unicode.org/~book/incoming/username -- do you have access to that? (Should be fine posting wherever is convenient.)

markusicu reviewed (approved):
looks plausible :-)

please squash & merge

Manishearth commented:
Done! Going to pull in the latest data soon

> We usually post to https://corp.unicode.org/~book/incoming/username -- do you have access to that? (Should be fine posting wherever is convenient.)

I have the password for `book`, but I don't know how to upload to it: I'm able to log in to the web with `book` and get a dir listing, but not to FTP as listed in https://corp.unicode.org/~book/bookurls.txt . Do i need to get an account?

Manishearth commented:
Posted https://github.com/unicode-org/unicodetools/pull/45 and https://corp.unicode.org/~book/incoming/manishearth/u14/d8/

Manishearth commented:
I made some doc changes to update https://sites.google.com/site/unicodetools/home to the latest set of steps.

macchiati commented:
Thanks!

Mark


On Tue, Mar 2, 2021 at 5:37 PM Manish Goregaokar <notifications@github.com>
wrote:

> I made some doc changes to update
> https://sites.google.com/site/unicodetools/home to the latest set of
> steps.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/unicode-org/unicodetools/pull/40#issuecomment-789358038>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACJLEMATLDAK4TJJWRHO45TTBWHGDANCNFSM4XOILQEA>
> .
>
